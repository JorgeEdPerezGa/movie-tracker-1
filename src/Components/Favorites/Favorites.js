import React, { Component } from 'react';
import { connect } from 'react-redux';
import Card from '../../Components/Card/Card';
import CardList from '../../Containers/CardList/CardList'
import { fetchFavorites, deleteFavorites, receiveFavorites } from '../../API/helper.js';
import { setFavorites } from '../../Actions/Actions';

class Favorites extends Component {


  addFavoriteMovie = (movie) => {
    if(this.props.user[0]) {
     this.checkFavorites(movie);
    } else {
      this.props.history.push('/login');
    }
  }

  checkFavorites = async (movie) => {
   const movieFoundInFavs = this.props.favorites.find(favorite => {
     return favorite.title === movie.title
   })
   if(!movieFoundInFavs) {
     this.toggleFavorite(movie)
     await fetchFavorites(movie, this.props.user[0].data.id);
     const newFavs = await receiveFavorites(this.props.user[0].data.id);
     return this.props.setFavorites(newFavs.data);
   } else {
     const deleteFavs = await deleteFavorites(movie, this.props.user[0].data.id )
     const newFavs = await receiveFavorites(this.props.user[0].data.id);
     return this.props.setFavorites(newFavs.data)
   }
  }

  toggleFavorite = (movie) => {
  movie.isFavorite = !movie.isFavorite;
}

  render() {
    const favoriteMovies = this.props.favorites.map((movie) => {
     return <Card
       key = {movie.id}
       movie={movie}
       title = {movie.title}
       overview = {movie.overview}
       poster = {movie.poster_path}
       vote = {movie.vote}
       backdrop = {movie.backdrop}
       addToFavorites={this.addFavoriteMovie}
       user={this.props.user}
       isFavorite={movie.isFavorite}
     />
   })

    return (
      <div className = 'card-list'>
       {favoriteMovies}
      </div>
    );
 }
}

const mapStateToProps = (store) => {
  return {
    user: store.user,
    movies: store.movies,
    favorites: store.favorite
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    setFavorites: (newFavorites) => {
      dispatch(setFavorites(newFavorites));
    }
  }
}


export default connect(mapStateToProps, mapDispatchToProps)(Favorites);
