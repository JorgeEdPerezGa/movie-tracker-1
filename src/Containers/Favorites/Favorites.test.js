import React from 'react';
import { shallow, mount } from 'enzyme';
import { Favorites, mapStateToProps, mapDispatchToProps } from './Favorites';

describe('favorites tests', () => {
  let favorites;
  beforeEach( () => {
    const mockFunc = jest.fn();
    favorites = shallow(
      <Favorites
        favorites={['array']}
        history={{hola: 'blah'}}
        location={{pathname: 'current-path'}}
        match={{hola: 'blah'}}
        movies={['coco']}
        setFavorites={mockFunc}
        user={['blah']}
      />);
    });

    it('should match the snapshot', () => {
      expect(favorites).toMatchSnapshot();
      });

    it('Should be defined', () => {
      expect(favorites).toBeDefined();
    })
  });

    describe('map state to props', () => {
      it('should receive movies from the store', () => {
        const mockStore = {
          movies: [{title: 'Star Wars'}]
        }
        const expected = mapStateToProps(mockStore)
        expect(expected.movies).toEqual(mockStore.movies)
      })

      describe('mapDispatchToProps', () => {
       let mockDispatch;
       let result;

       beforeEach(() => {
         mockDispatch = jest.fn();
         result = mapDispatchToProps(mockDispatch);
       });

       it('should call dispatch when setFavorites is called', () => {
         result.setFavorites();
         expect(mockDispatch).toHaveBeenCalled();
       });

       it('should call dispatch when removeFav is called', () => {
         result.setFavorites();
         expect(mockDispatch).toHaveBeenCalled();
       });
     });
  })
