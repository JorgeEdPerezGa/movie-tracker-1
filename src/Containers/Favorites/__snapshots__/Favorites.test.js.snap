// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`favorites tests should match the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Favorites
    favorites={
      Array [
        "array",
      ]
    }
    history={
      Object {
        "hola": "blah",
      }
    }
    location={
      Object {
        "pathname": "current-path",
      }
    }
    match={
      Object {
        "hola": "blah",
      }
    }
    movies={
      Array [
        "coco",
      ]
    }
    setFavorites={[Function]}
    user={
      Array [
        "blah",
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Card
          addToFavorites={[Function]}
          backdrop={undefined}
          isFavorite={undefined}
          movie="array"
          overview={undefined}
          poster={undefined}
          title={undefined}
          user={
            Array [
              "blah",
            ]
          }
          vote={undefined}
        />,
      ],
      "className": "card-list",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "addToFavorites": [Function],
          "backdrop": undefined,
          "isFavorite": undefined,
          "movie": "array",
          "overview": undefined,
          "poster": undefined,
          "title": undefined,
          "user": Array [
            "blah",
          ],
          "vote": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Card
            addToFavorites={[Function]}
            backdrop={undefined}
            isFavorite={undefined}
            movie="array"
            overview={undefined}
            poster={undefined}
            title={undefined}
            user={
              Array [
                "blah",
              ]
            }
            vote={undefined}
          />,
        ],
        "className": "card-list",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "addToFavorites": [Function],
            "backdrop": undefined,
            "isFavorite": undefined,
            "movie": "array",
            "overview": undefined,
            "poster": undefined,
            "title": undefined,
            "user": Array [
              "blah",
            ],
            "vote": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
